File: ios/Classes/LibvncviewerFlutterPlugin.m
#import "LibvncviewerFlutterPlugin.h"
#import "VncClient.h"
#import "GLRender.h"
static NSMutableDictionary *flutterEventSinkDictionary = nil;

static void rfbClientCallback(int64_t id,int code,NSString* flag,NSString* msg){
    dispatch_async(dispatch_get_main_queue(), ^{
        FlutterEventSink sink = [flutterEventSinkDictionary objectForKey:@(id)];
        if(sink){
            NSMutableDictionary *mutableDictionary = [NSMutableDictionary dictionary];
            [mutableDictionary setObject:flag forKey:@"flag"];
            [mutableDictionary setObject:@(code) forKey:@"code"];
            [mutableDictionary setObject:msg forKey:@"msg"];
            sink(mutableDictionary);
        }
    });
}

@interface LibvncviewerFlutterPlugin()<FlutterStreamHandler>

@property (nonatomic, strong) NSObject<FlutterTextureRegistry> *textures;

@property (nonatomic, strong) GLRender *glRender;

@end

@implementation LibvncviewerFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(!flutterEventSinkDictionary){
        flutterEventSinkDictionary = [[NSMutableDictionary alloc]init];
    }
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"libvncviewer_flutter"
                                     binaryMessenger:[registrar messenger]];
    LibvncviewerFlutterPlugin* instance = [[LibvncviewerFlutterPlugin alloc] init];
    instance.textures=registrar.textures;
    [registrar addMethodCallDelegate:instance channel:channel];
    
    FlutterEventChannel* eventchannel = [FlutterEventChannel eventChannelWithName:@"libvncviewer_flutter_eventchannel" binaryMessenger:[registrar messenger]];
    [eventchannel setStreamHandler:instance];
}

#pragma mark - <FlutterStreamHandler>
- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(FlutterEventSink)events {
    int64_t clientId=[[arguments objectForKey:@"clientId"] longLongValue];
    [flutterEventSinkDictionary setObject:events forKey:@(clientId)];
    events(@{@"flag":@"onReady"});
    return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments{
    int64_t clientId=[[arguments objectForKey:@"clientId"] longLongValue];
    [flutterEventSinkDictionary removeObjectForKey:@(clientId)];
    return nil;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"getPlatformVersion" isEqualToString:call.method]) {
        result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    }
    if ([@"closeVncClient" isEqualToString:call.method]) {
        int64_t clientId=[[call.arguments objectForKey:@"clientId"] longLongValue];
        [[VncClient getVncClient:clientId] close];
    }
    if ([@"startVncClient" isEqualToString:call.method]) {
        int64_t clientId = [[call.arguments objectForKey:@"clientId"] longLongValue];
        [[VncClient getVncClient:clientId] connect];
    }
    if ([@"sendPointer" isEqualToString:call.method]) {
        int64_t clientId=[[call.arguments objectForKey:@"clientId"] longLongValue];
        int x=[[call.arguments objectForKey:@"x"] intValue];
        int y=[[call.arguments objectForKey:@"y"] intValue];
        int mask=[[call.arguments objectForKey:@"mask"] intValue];
        [[VncClient getVncClient:clientId] sendPointer:x andY:y andButtonMask:mask];
    }
    if ([@"sendKey" isEqualToString:call.method]) {
        int64_t clientId=[[call.arguments objectForKey:@"clientId"] longLongValue];
        int key=[[call.arguments objectForKey:@"key"] intValue];
        int down=[[call.arguments objectForKey:@"down"] intValue];
        [[VncClient getVncClient:clientId] sendKey:key andDown:down];
    }

    if ([@"initVncClient" isEqualToString:call.method]) {
        NSString* hostName=[call.arguments objectForKey:@"hostName"];
        int port=[[call.arguments objectForKey:@"port"] intValue];
        NSString* password=[call.arguments objectForKey:@"password"];
        VncClient* client = [[VncClient alloc] initWith:hostName andPort:port andPassword:password];
        
        __weak typeof(self) wself = self;
        
        [client registerInfoCallBack:rfbClientCallback];
        
        [client registerFrameCallBack:^(uint8_t*data,int w,int h){
            dispatch_async(dispatch_get_main_queue(), ^{
                [wself.textures textureFrameAvailable:client.textureId];
            });
        }];
        
        [client registerImageResizeCallBack:^(int width, int height) {
            client.textureId = [self.textures registerTexture:[client getGLRender]];
            dispatch_async(dispatch_get_main_queue(), ^{
                FlutterEventSink sink = [flutterEventSinkDictionary objectForKey:@(client.getClientId)];
                if(sink){
                    sink(@{@"flag":@"imageResize",@"width":@(width),@"height":@(height),@"textureId":@(client.textureId)});
                }
            });
            
        }];
        
        [client initRfbClient];
        
        result(@([client getClientId]));
    }
    //    result(FlutterMethodNotImplemented);
}

@end

File: macos/Classes/LibvncviewerFlutterPlugin.swift
import Cocoa
import FlutterMacOS

public class LibvncviewerFlutterPlugin: NSObject, FlutterPlugin,FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        let params = arguments as! NSDictionary
        let clientId = params.object(forKey: "clientId") as! Int64
        clientEventChannel[clientId]=events
        events(["flag":"onReady"])
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        let params = arguments as! NSDictionary
        let clientId = params.object(forKey: "clientId") as! Int64
        clientEventChannel.removeValue(forKey: clientId)
        return nil
    }
    
    
    private var textureRegistry: FlutterTextureRegistry?
    
    private var clientEventChannel:[Int64:FlutterEventSink] = [:]
    
    private var renderer: [Int64: MyTexture] = [:]
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "libvncviewer_flutter", binaryMessenger: registrar.messenger)
        let eventChannel = FlutterEventChannel(name: "libvncviewer_flutter_eventchannel", binaryMessenger: registrar.messenger)
        let instance = LibvncviewerFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
        instance.textureRegistry=registrar.textures
        
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)
        case "initVncClient":
            let params = call.arguments as! NSDictionary
            let hostName = params.object(forKey: "hostName") as! String
            let port = params.object(forKey: "port") as! Int32
            let password = params.object(forKey: "password") as! String
            let client = VncClient(hostName, andPort: port, andPassword: password)
            client.registerImageResizeCallBack {  width, height in
                let myTexture = MyTexture(width: Int(width), height: Int(height))
                let textureId = self.textureRegistry?.register(myTexture) ?? -1
                client.textureId=textureId
                self.renderer[client.getId()]=myTexture
                let sink = self.clientEventChannel[client.getId()]!
                sink(["flag":"imageResize","width":width,"height":height,"textureId":textureId]);
            }
            client.registerInfoCallBack { clientId, code, flag, msg in
                let sink = self.clientEventChannel[clientId]!
                sink(["flag":flag,"code":code,"msg":msg]);
            }
            client.registerFrameCallBack { data, w, h in
                self.renderer[client.getId()]?.markFrameAvaliableRaw(buffer: data, len: Int(client.getFrameBufferSize()), width: Int(w), height: Int(h), stride_align: 1)
                self.textureRegistry?.textureFrameAvailable(client.textureId)
                
            }
            client.initRfbClient()
            result(client.getId())
        case "startVncClient":
            let params = call.arguments as! NSDictionary
            let clientId = params.object(forKey: "clientId") as! Int64
            VncClient.getVncClient(clientId).connect()
        case "closeVncClient":
            let params = call.arguments as! NSDictionary
            let clientId = params.object(forKey: "clientId") as! Int64
            let client = VncClient.getVncClient(clientId)
            client.close()
            self.textureRegistry?.unregisterTexture(client.textureId)
            self.renderer.removeValue(forKey: clientId)
        case "sendPointer":
            let params = call.arguments as! NSDictionary
            let clientId = params.object(forKey: "clientId") as! Int64
            let x = params.object(forKey: "x") as! Int32
            let y = params.object(forKey: "y") as! Int32
            let mask = params.object(forKey: "mask") as! Int32
            VncClient.getVncClient(clientId).sendPointer(x, andY: y, andButtonMask: mask)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
